File: platforms/ios/runtime/Sources/NativeblocksRuntime/frame/BlockTree.swift
import NativeblocksRuntimeFFI
import SwiftUI

internal struct NativeFrame: View {

    private let route: String
    private let args: [String: String]
    private let state: NativeblocksFrameState
    private let loading: () -> AnyView
    private let error: (String) -> AnyView
    private let instanceName: String

    @StateObject private var frameViewModel: FrameViewModel
    @Environment(\.verticalSizeClass) private var verticalSizeClass
    @Environment(\.horizontalSizeClass) private var horizontalSizeClass

    init(
        instanceName: String = "default",
        route: String,
        args: [String: String],
        state: NativeblocksFrameState = .stateless,
        loading: @escaping () -> AnyView,
        error: @escaping (String) -> AnyView
    ) {
        self.instanceName = instanceName
        self.route = route
        self.args = args
        self.state = state
        self.loading = loading
        self.error = error
        self._frameViewModel = StateObject(
            wrappedValue: NativeCoreSDKInjector.get(name: instanceName).makeFrameViewModel()
        )
    }

    var body: some View {
        Group {
            switch frameViewModel.renderingState {
            case .loading:
                loading()
            case .ready:
                RootLifecycle(instanceName: instanceName, vm: frameViewModel)
            case .error(let message):
                error(message)
            }
        }
        .task(id: FrameSetup(instanceName: instanceName, route: route, args: args, stateKey: state.key)) {
            frameViewModel.setupFrame(route: route, args: args, stateKey: state.key)
        }
        .onDisappear {
            frameViewModel.releaseFrame()
        }
    }
}

private struct FrameSetup: Equatable {
    let instanceName: String
    let route: String
    let args: [String: String]
    let stateKey: String?
}

private struct RootLifecycle: View {

    let instanceName: String
    @ObservedObject var vm: FrameViewModel

    var body: some View {
        Group {
            if let rootKey = vm.rootKey {
                Block(
                    instanceName: instanceName,
                    vm: vm,
                    blockKey: rootKey,
                    listItemIndex: NONE_INDEX,
                    parentScope: nil,
                    resolver: nil
                )
                    .task(id: vm.frameUpdateGeneration) {
                        vm.rootEntered(rootKey)
                    }
                    .onDisappear {
                        vm.rootExited(rootKey)
                    }
            } else {
                EmptyView()
            }
        }
    }
}

private struct Block: View {

    let instanceName: String
    @ObservedObject var vm: FrameViewModel
    let blockKey: String
    let listItemIndex: Int
    let parentScope: Any?
    var resolver: TemplateResolver? = nil

    var body: some View {
        if let block = vm.blockOf(blockKey), vm.valueOf(block.visibility) != "false" {
            let blockContext = makeBlockContext(for: block)
            if block.keyType == "ROOT" {
                AnyView(RootBlock(blockContext: blockContext))
            } else if case .rendering(let render) = vm.blockProvider.getProvidedBlocks()[block.keyType] {
                AnyView(render(blockContext))
            } else if let fallbackBlock = vm.blockProvider.getFallbackBlock() {
                AnyView(fallbackBlock(block.keyType, blockKey))
                    .onAppear { vm.logBlockFallback(keyType: block.keyType, blockKey: blockKey) }
            } else {
                InternalFallbackBlock(key: block.keyType)
                    .onAppear { vm.logBlockFallback(keyType: block.keyType, blockKey: blockKey) }
            }
        } else {
            EmptyView()
        }
    }

    private func blockModifier(for block: NativeBlockModel) -> NativeblocksModifier {
        let items = block.modifiers
        if items.isEmpty {
            return .none
        }
        let provided = vm.modifierProvider.getProvidedModifiers()
        let instanceName = instanceName
        let listItemIndex = listItemIndex
        let parentScope = parentScope
        let resolver = resolver
        let vm = vm
        return NativeblocksModifier { content in
            var view = content
            for item in items {
                if let nativeModifier = provided[item.keyType] {
                    let modifierContext = ModifierContext(
                        instanceName: instanceName,
                        listItemIndex: listItemIndex,
                        onFindVariable: { data in
                            vm.valueOf(data?.value ?? "")
                        },
                        onUpdateVariable: { data, value in
                            guard let data = data else { return }
                            vm.updateVariable(key: data.value, value: value)
                        },
                        onFindAction: { eventType in
                            vm.actionOf(blockKey: blockKey, eventType: eventType)
                        },
                        onHandleAction: { index, action, event in
                            vm.handleAction(index, action, event, parentScope)
                        },
                        modifier: item,
                        scope: parentScope,
                        resolveTemplate: { resolver?.resolve($0) ?? $0 }
                    )
                    view = nativeModifier(view, modifierContext)
                } else {
                    vm.logModifierFallback(keyType: item.keyType, blockKey: blockKey)
                }
            }
            return view
        }
    }

    private func makeBlockContext(for block: NativeBlockModel, resolver: TemplateResolver? = nil) -> BlockContext {
        let resolver = resolver ?? self.resolver
        return BlockContext(
            instanceName: instanceName,
            listItemIndex: listItemIndex,
            onFindVariable: { data in
                vm.valueOf(data?.value ?? "")
            },
            onUpdateVariable: { data, value in
                guard let data = data else { return }
                vm.updateVariable(key: data.value, value: value)
            },
            onFindAction: { eventType in
                vm.actionOf(blockKey: blockKey, eventType: eventType)
            },
            onHandleAction: { index, action, event in
                vm.handleAction(index, action, event, parentScope)
            },
            block: block,
            modifier: blockModifier(for: block),
            onDescribeSubBlock: { blockKeys, subSlot, describeScope in
                return AnyView(
                    ForEach(blockKeys[subSlot.slot] ?? [], id: \.self) { childKey in
                        if let child = vm.blockOf(childKey),
                            case .describing(let describe) = vm.blockProvider.getProvidedBlocks()[child.keyType] {
                            AnyView(
                                describe(
                                    makeBlockContext(
                                        for: child,
                                        resolver: (describeScope as? TemplateResolver) ?? resolver
                                    ),
                                    describeScope
                                )
                            )
                        } else {
                            Color.clear.frame(width: 0, height: 0)
                                .onAppear { vm.logBlockFallback(keyType: block.keyType, blockKey: childKey) }
                        }
                    }
                )
            },
            onSubBlock: { blockKeys, subSlot, itemIndex, scope in
                return AnyView(
                    ForEach(blockKeys[subSlot.slot] ?? [], id: \.self) { childKey in
                        Block(
                            instanceName: instanceName,
                            vm: vm,
                            blockKey: childKey,
                            listItemIndex: itemIndex == NONE_INDEX ? listItemIndex : itemIndex,
                            parentScope: scope,
                            resolver: (scope as? TemplateResolver) ?? resolver
                        )
                    }
                )
            },
            scope: parentScope,
            resolveTemplate: { resolver?.resolve($0) ?? $0 }
        )
    }
}

File: platforms/android/runtime/core/src/main/java/io/nativeblocks/runtime/frame/BlockTree.kt
@file:Suppress("DEPRECATION")

package io.nativeblocks.runtime.frame

import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.key
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import io.nativeblocks.runtime.api.provider.block.BlockContext
import io.nativeblocks.runtime.api.provider.block.NONE_INDEX
import io.nativeblocks.runtime.api.provider.block.defaults.InternalFallbackBlock
import io.nativeblocks.runtime.api.provider.block.defaults.RootBlock
import io.nativeblocks.runtime.api.provider.model.NativeBlockModel
import io.nativeblocks.runtime.api.provider.modifier.ModifierContext
import io.nativeblocks.runtime.api.util.LocalNativeWindowWidthClass
import io.nativeblocks.runtime.api.util.currentWindowWidthClass
import io.nativeblocks.runtime.ffi.RenderingState

@Composable
internal fun NativeFrame(
    instanceName: String,
    frameViewModel: FrameViewModel,
    loading: @Composable () -> Unit,
    error: @Composable (String) -> Unit
) {
    val state by frameViewModel.renderingState.collectAsStateWithLifecycle()
    CompositionLocalProvider(LocalNativeWindowWidthClass provides currentWindowWidthClass()) {
        when (val frameState = state) {
            is RenderingState.Loading -> loading.invoke()
            is RenderingState.Ready -> RootLifecycle(instanceName, frameViewModel)
            is RenderingState.Error -> error.invoke(frameState.message)
        }
    }
}

@Composable
private fun RootLifecycle(instanceName: String, vm: FrameViewModel) {
    val rootKeyState by vm.rootKey.collectAsStateWithLifecycle()
    val appearGen by vm.frameUpdateGeneration.collectAsStateWithLifecycle()

    val rootKey = rootKeyState ?: return

    DisposableEffect(rootKey) {
        onDispose { vm.rootExited(rootKey) }
    }

    LaunchedEffect(appearGen) {
        vm.rootEntered(rootKey)
    }

    Block(instanceName, vm, rootKey, NONE_INDEX, appearGen, parentScope = null)
}

@Composable
private fun Block(
    instanceName: String,
    vm: FrameViewModel,
    blockKey: String,
    listItemIndex: Int,
    frameGeneration: Int,
    parentScope: Any?,
) {
    val block = vm.blockOf(blockKey) ?: return
    if (vm.valueOf(block.visibility) == "false") return

    val nativeBlock = remember(block.keyType) {
        if (block.keyType == "ROOT") { blockContext -> RootBlock(blockContext) }
        else vm.blockProvider.getProvidedBlocks()[block.keyType]
    }
    if (nativeBlock == null) {
        vm.logBlockFallback(block.keyType, blockKey)
        vm.blockProvider.getFallbackBlock()?.invoke(block.keyType, blockKey) ?: InternalFallbackBlock(block.keyType)
        return
    }

    val modifier = blockModifier(instanceName, vm, block, blockKey, listItemIndex, parentScope)

    val blockContext = remember(block, listItemIndex, modifier) {
        BlockContext(
            instanceName = instanceName,
            listItemIndex = listItemIndex,
            onFindVariable = { data -> vm.valueOf(data?.value.orEmpty()) },
            onUpdateVariable = { data, value -> vm.updateVariable(data?.value.orEmpty(), value) },
            onFindAction = { vm.actionOf(blockKey, it) },
            onHandleAction = { index, action, event -> vm.handleAction(index, action, event) },
            block = block,
            modifier = modifier,
            onSubBlock = { blockKeys, subSlot, itemIndex, scope ->
                blockKeys[subSlot.slot]?.forEach { childKey ->
                    key(childKey) {
                        Block(
                            instanceName = instanceName,
                            vm = vm,
                            blockKey = childKey,
                            listItemIndex = if (itemIndex == NONE_INDEX) listItemIndex else itemIndex,
                            frameGeneration = frameGeneration,
                            parentScope = scope,
                        )
                    }
                }
            },
        )
    }

    nativeBlock.invoke(blockContext)
}

@Composable
private fun blockModifier(
    instanceName: String,
    vm: FrameViewModel,
    block: NativeBlockModel,
    blockKey: String,
    listItemIndex: Int,
    parentScope: Any?,
): Modifier {
    if (block.modifiers.isEmpty()) {
        return Modifier
    }
    val provided = vm.modifierProvider.getProvidedModifiers()
    var chain: Modifier = Modifier
    block.modifiers.forEach { item ->
        val nativeModifier = provided[item.keyType]
        if (nativeModifier != null) {
            val modifierContext = ModifierContext(
                instanceName = instanceName,
                listItemIndex = listItemIndex,
                onFindVariable = { data -> vm.valueOf(data?.value.orEmpty()) },
                onUpdateVariable = { data, value -> vm.updateVariable(data?.value.orEmpty(), value) },
                onFindAction = { vm.actionOf(blockKey, it) },
                onHandleAction = { index, action, event -> vm.handleAction(index, action, event) },
                modifier = item,
                scope = parentScope,
            )
            chain = chain.then(key(item.keyType, item.position) { nativeModifier.invoke(modifierContext) })
        } else {
            vm.logModifierFallback(item.keyType, blockKey)
        }
    }
    return chain
}
